File: Core-backend-layer/Programming-fundamentals/Java/OOP/src/Encapsulation/Product.java
import java.time.LocalDateTime;

public class Product {
    // Private fields
    private String name;
    private double price;
    private final int productId;
    private String manufacturer;
    private LocalDateTime productionDate;

    // Constructor
    public Product(String name, double price, int productId, String manufacturer, LocalDateTime productionDate) {
        setName(name);
        setPrice(price);
        this.productId = productId;
        this.manufacturer = manufacturer;
        setProductionDate(productionDate);
    }

    // Getter methods
    public String getName() {
        return name;
    }

    public double getPrice() {
        return price;
    }

    public int getProductId() {
        return productId;
    }

    public String getManufacturer() {
        return manufacturer;
    }

    public LocalDateTime getProductionDate() {
        return productionDate;
    }

    // Setter methods with validation
    public void setName(String name) {
        if (name != null && !name.trim().isEmpty()) {
            this.name = name;
        } else {
            System.out.println("Invalid name");
        }
    }

    public void setPrice(double price) {
        if (price >= 0) {
            this.price = price;
        } else {
            System.out.println("Invalid price. Price cannot be negative.");
        }
    }

    public void setManufacturer(String manufacturer) {
        this.manufacturer = manufacturer;
    }

    public void setProductionDate(LocalDateTime productionDate) {
        if (productionDate != null && productionDate.isBefore(LocalDateTime.now())) {
            this.productionDate = productionDate;
        } else {
            System.out.println("Invalid production date. Production date cannot be in the future.");
        }
    }

    public void displayInfo() {
        System.out.println("Product ID: " + productId + ", Product: " + name + ", Price: $" + price
                + ", Manufacturer: " + (manufacturer != null ? manufacturer : "N/A")
                + ", Production date: " + (productionDate != null ? productionDate : "N/A"));
    }
}

File: Core-backend-layer/Programming-fundamentals/Python/OOP/src/encapsulation/product.py
from datetime import datetime

class Product:
    def __init__(self, name, price, product_id, manufacturer, production_date):
        self._name = None  # Private field with leading underscore
        self._price = 0.0
        self._product_id = product_id
        self._manufacturer = None
        self._production_date = None
        # Use setter methods to enforce validation
        self.set_name(name)
        self.set_price(price)
        self.set_manufacturer(manufacturer)
        self.set_production_date(production_date)

    def get_name(self):
        return self.name

    def get_price(self):
        return self.price

    def get_product_id(self):
        return self.product_id

    def get_manufacturer(self):
        return self.manufacturer

    def get_production_date(self):
        return self.production_date

    def set_name(self, name):
        if name is not None and name.strip():
            self._name = name
        else:
            print("Invalid name")

    def set_price(self, price):
        if price >= 0:
            self._price = price
        else:
            print("Invalid price. Price cannot be negative.")

    def set_manufacturer(self, manufacturer):
        self._manufacturer = manufacturer

    def set_production_date(self, production_date):
        if production_date is not None and production_date < datetime.now():
            self._production_date = production_date
        else:
            print("Invalid production date. Production date cannot be in the future.")

    def display_info(self):
        print(f"Product ID: {self._product_id}, Product: {self._name}, Price: ${self._price}, "
              f"Manufacturer: {self._manufacturer if self._manufacturer is not None else 'N/A'}, "
              f"Production date: {self._production_date if self._production_date is not None else 'N/A'}")